src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java

```diff
@@ -345,40 +346,45 @@ public static synchronized boolean addEventToJenkinsJobRun(Build build)
         return false;
     }
 
-    public static synchronized void addBuildToNoBCList(Build build) {
+    private static void addBuildToNoBCList(Build build) {
         // should have been caught upstack, but just in case since public method
         if (!OpenShiftUtils.isPipelineStrategyBuild(build))
             return;
-        buildsWithNoBCList.add(build);
-    }
-
-    private static synchronized void removeBuildFromNoBCList(Build build) {
-        buildsWithNoBCList.remove(build);
+        try {
+          buildsWithNoBCList.add(build);
+        } catch (ConcurrentModificationException | IllegalArgumentException |
+          UnsupportedOperationException | NullPointerException e) {
+          logger.log(Level.WARNING,"Failed to add item " +
+            build.getMetadata().getName(), e);
+        }
     }
 
-    private static synchronized void clearNoBCList() {
-        buildsWithNoBCList.clear();
+    private static void removeBuildFromNoBCList(Build build) {
+          buildsWithNoBCList.remove(build);
     }
 
     // trigger any builds whose watch events arrived before the
     // corresponding build config watch events
-    public static synchronized void flushBuildsWithNoBCList() {
-        HashSet<Build> clone = (HashSet<Build>) buildsWithNoBCList.clone();
-        clearNoBCList();
-        for (Build build : clone) {
-            WorkflowJob job = getJobFromBuild(build);
-            if (job != null)
-                try {
-                    logger.info("triggering job run for previously skipped build "
-                            + build.getMetadata().getName());
-                    triggerJob(job, build);
-                } catch (IOException e) {
-                    logger.log(Level.WARNING, "flushCachedBuilds", e);
-                }
-            else
-                addBuildToNoBCList(build);
+    public static void flushBuildsWithNoBCList() {
+        for (Build build : buildsWithNoBCList) {
+          WorkflowJob job = getJobFromBuild(build);
+          if (job != null) {
+            try {
+              logger.info("triggering job run for previously skipped build "
+                + build.getMetadata().getName());
+              triggerJob(job, build);
+            } catch (IOException e) {
+              logger.log(Level.WARNING, "flushCachedBuilds", e);
+            }
+            try {
+              removeBuildFromNoBCList(build);
+            } catch (UnsupportedOperationException | IllegalStateException |
```

Comment:
If you were getting problems with `iter.remove` now, like unsupported operation, with the `ConcurrentHashSet` then we need to rethink this

we may be getting lucky with you not hitting concurrent mod exceptions, but that will impact CPU if it ever did

Reply:
well, I use those exceptions as they part of the Concurrent in particularly, instead of usign general Exception.

also, we may want to log any concurrent mod exceptions, we faced it somehow

Reply:
No need re: concurrent mod ... see my comment above

Since your are logging the exception, rather than only catching specific exceptions, catch Throwable ... I don't want any exceptions to escape here